lib/cn/csched_sp3_work.c

```diff
@@ -278,25 +285,98 @@ sp3_work_wtype_idle(
     return 0;
 }
 
+/* This function is periodically invoked for each node that needs to be
+ * split until the function returns non-zero.
+ *
+ * If conditions are favorable to split we set the "tn_splitting" flag
+ * to request a split.  Until then, we must make the following checks
+ * on every invocation:
+ *
+ * 1) If the root node is too long then wait for root spill to catch up.
+ *
+ * 2) If there's an active spill to this node then defer requesting a split
+ *    in hopes we can split some other node that isn't currently undergoing
+ *    an active spill.  We try this at most a few times, because if the tree
+ *    has "split_cnt_max" nodes or fewer they might all be undergoing an
+ *    active spill.
+ *
+ * 3) Defer requesting a split if there are already "split_cnt_max" splits
+ *    active in the tree (i.e., the max number of concurrent splits).
+ *
+ * 4) Defer requesting a split if sufficient time has not passed since
+ *    the first of the last batch of splits started.
+ *
+ * If none of the above conditions hold, then we request a split by setting
+ * the "tn_splitting" flag to true, which will prevent new spills into this
+ * node.  However, if this node is currently undergoing an active spill then
+ * we must return 0 to avoid starting a split.  We then re-evalutate this
+ * condition on each invocation until all spills to this node have completed
+ * (i.e., tn_ss_spilling == 0).  Only then may we start the split.
+ *
+ * The reason for these machinations is to avoid stalling the root spill
+ * pipeline, as that adversely impacts the ingest pipeline, which is
+ * then seen as poor quality-of-service from the client's perspective.
```

Comment:
That's the most important part of this comment.   Put the comment *before* the list of "machinations".

Reply:
Oops, I forgot to update this note, it's way out of data...
Haha, will reorganize..